### PR TITLE
Fixing 1163

### PR DESCRIPTION
**Pull request by hkaiser**
This proposes two new components: `unordered_map` and `component_storage` (sorry I was not able to keep those separate...)

 - `unordered_map` is a distributed hash-map implementation aligned with `std::unordered_map`. It currently exposes a minimalistic interface (no iterators). Even so it's already useful and is used for the component_storage module. We will also apply it to implement Intel CnC on top of HPX.

 - `component_storage` is a simple module which supports migrating component instances out of main memory and store it as a byte stream. These instances can be migrated back into main memory and will retain their global id. This is the first implementation of functionality needed for the PXFS project.

### CODE REVIEW

**Comment by hkaiser** [comment]
FWIW, all tests are green for this branch.

File: hpx/components/containers/unordered/partition_unordered_map_component.hpp [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+        /// Return the element at the position \a pos in the partition_unordered_map
+        /// container.
+        ///
+        /// \param pos Positions of the elements in the partition_unordered_map
+        ///
+        /// \return Return the values of the elements at position represented
+        ///         by \a pos.
+        ///
+        std::vector<T> get_values(std::vector<Key> const& keys)
+        {
```
Comment (sithhell) [inline hpx/components/containers/unordered/partition_unordered_map_component.hpp]: This function is missing a check that the keys are really contained in this partitions

File: hpx/components/containers/unordered/partition_unordered_map_component.hpp [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+
+        /// Copy the value of \a val in the element at position \a pos in the
+        /// partition_unordered_map container.
+        ///
+        /// \param pos   Position of the element in the partition_unordered_map
+        ///
+        /// \param val   The value to be copied
+        ///
+        void set_value(Key const& pos, T const& val)
+        {
```
Comment (sithhell) [inline hpx/components/containers/unordered/partition_unordered_map_component.hpp]: Would a sanity check that we ended up in the right partition make sense here?

File: hpx/components/containers/unordered/partition_unordered_map_component.hpp [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+        /// Copy the value of \a val for the elements at positions \a pos in
+        /// the partition_unordered_map container.
+        ///
+        /// \param pos   Positions of the elements in the partition_unordered_map
+        ///
+        /// \param val   The value to be copied
+        ///
+        void set_values(std::vector<Key> const& keys,
+            std::vector<T> const& val)
+        {
```
Comment (sithhell) [inline hpx/components/containers/unordered/partition_unordered_map_component.hpp]: Same as above.

File: hpx/components/containers/unordered/partition_unordered_map_component.hpp [hunk 4]
```diff
@@ -0,0 +1,601 @@
...
+        /// partition_unordered_map with size 0.
+        ///
+        void clear()
+        {
+            partition_unordered_map_.clear();
+        }
+
+        /// Erase the given element
+        std::size_t erase(Key const& key)
+        {
```
Comment (sithhell) [inline hpx/components/containers/unordered/partition_unordered_map_component.hpp]: Check if key is really part of the partition is missing.

File: hpx/components/containers/unordered/unordered_map.hpp [hunk 5]
```diff
@@ -0,0 +1,932 @@
...
+
+        // This is the vector representing the base_index and corresponding
+        // global ID's of the underlying partition_vectors.
+        partitions_vector_type partitions_;
+
+        // will be set for created (non-attached) objects
+        std::string registered_name_;
+
+        ///////////////////////////////////////////////////////////////////////
+        // Connect this vector to the existing vector using the given symbolic
```
Comment (sithhell) [inline hpx/components/containers/unordered/unordered_map.hpp]: copy&paste error: vector->unordered map

File: tests/unit/components/migrate_component_to_storage.cpp [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        HPX_TEST_EQ(storage.size_sync(), std::size_t(1));
+
+        // make sure all references go out of scope
+    }
+
+    HPX_TEST_EQ(storage.size_sync(), std::size_t(1));
+
+    {
+        test_client t1(hpx::components::migrate_from_storage<test_server>(
+            oldid, storage));
```
Comment (sithhell) [inline tests/unit/components/migrate_component_to_storage.cpp]: Why do we need the old id there? Shouldn't the storage component transparently take care that the GID is preserved?

**Comment by sithhell** [comment]
In general looks good to me. What I miss, and what I couldn't spot right away from studying the code is how to construct a unordered map or container in true SPMD style, same goes for the component storage.
A few other questions:
- What happens when I call an action on a component that has been migrated to storage?
- Is it planned to support the choosing of the storage, like local disk/tape/whatever?

File: tests/unit/components/migrate_component_to_storage.cpp (hunk 6 above)
Comment (hkaiser) [inline tests/unit/components/migrate_component_to_storage.cpp]: We need it to extract the component form the storage.

File: hpx/components/containers/unordered/unordered_map.hpp (hunk 5 above)
Comment (hkaiser) [inline hpx/components/containers/unordered/unordered_map.hpp]: Good catch

File: hpx/components/containers/unordered/partition_unordered_map_component.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/components/containers/unordered/partition_unordered_map_component.hpp]: Nod, will add. Thanks!

File: hpx/components/containers/unordered/partition_unordered_map_component.hpp (hunk 2 above)
Comment (hkaiser) [inline hpx/components/containers/unordered/partition_unordered_map_component.hpp]: Not sure if this is possible. In order to calculate the sequence umber of the current partition we would need to know how many partitions are created overall. This information is not available to each of the partitions themselves (at least not now).

File: hpx/components/containers/unordered/partition_unordered_map_component.hpp (hunk 4 above)
Comment (hkaiser) [inline hpx/components/containers/unordered/partition_unordered_map_component.hpp]: The function returns the number of deleted elements. No need to check whether the key exists in the container.

**Comment by hkaiser** [comment]
> What I miss, and what I couldn't spot right away from studying the code is how to construct a 
> unordered map or container in true SPMD style, same goes for the component storage.

This is orthogonal to this PR, I'll try to create an example to demonstrate this, though.

> A few other questions:
>  - What happens when I call an action on a component that has been migrated to storage?

At this point it will blow up. Later on I would like to transparently bring it back before applying the action.

>  - Is it planned to support the choosing of the storage, like local disk/tape/whatever?

This would be a different storage type. Once we have that the `migrate_to_strage()` and `migrate_from_storage()` functions can be made generic.

**Comment by hkaiser** [comment]
Sorry, closed it accidentally...

**Comment by sithhell** [comment]
Yes, the points I raised in my previous comment are orthogonal to this PR. Those were just about further directions and general possibilities.

File: tests/unit/components/migrate_component_to_storage.cpp (hunk 6 above)
Comment (sithhell) [inline tests/unit/components/migrate_component_to_storage.cpp]: Hmm, good point. It looks strange at first sight, but I guess there is no way around it, unless you want to restore all stored components with that type, which might not really be what you want/need.

File: hpx/components/containers/unordered/partition_unordered_map_component.hpp (hunk 4 above)
Comment (sithhell) [inline hpx/components/containers/unordered/partition_unordered_map_component.hpp]: Right, good point.

**Pull request merged**